components/brave_shields/adblock/rs/src/engine.rs

```diff
@@ -0,0 +1,202 @@
+}
+
+pub fn engine_with_rules(rules: &CxxVector<u8>) -> BoxEngineResult {
+    || -> Result<Box<Engine>, InternalError> {
+        let mut filter_set = FilterSet::new(false);
+        filter_set.add_filter_list(std::str::from_utf8(rules.as_slice())?, Default::default());
+        let engine = InnerEngine::from_filter_set(filter_set, true);
+        Ok(Box::new(Engine { engine }))
+    }()
+    .into()
+}
+
+struct DomainResolver;
+
+impl ResolvesDomain for DomainResolver {
+    fn get_host_domain(&self, host: &str) -> (usize, usize) {
+        let_cxx_string!(host_cxx_string = host);
+        let position = resolve_domain_position(&host_cxx_string);
+        (position.start as usize, position.end as usize)
+    }
+}
+
+pub fn set_domain_resolver() -> bool {
+    adblock::url_parser::set_domain_resolver(Box::new(DomainResolver)).is_ok()
+}
+
+pub fn read_list_metadata(list: &CxxVector<u8>) -> FilterListMetadataResult {
+    || -> Result<FilterListMetadata, InternalError> {
+        Ok(adblock::lists::read_list_metadata(std::str::from_utf8(list.as_slice())?).into())
+    }()
+    .into()
+}
+
+fn convert_cxx_string_vector_to_string_collection<C>(
+    value: &CxxVector<CxxString>,
+) -> Result<C, Utf8Error>
+where
+    C: FromIterator<String>,
+{
+    value.iter().map(|s| s.to_str().map(|t| t.to_string())).collect()
+}
+
+impl Engine {
+    pub fn enable_tag(&mut self, tag: &CxxString) -> EmptyTupleResult {
+        || -> Result<(), InternalError> { Ok(self.engine.enable_tags(&[tag.to_str()?])) }().into()
+    }
+
+    pub fn disable_tag(&mut self, tag: &CxxString) -> EmptyTupleResult {
+        || -> Result<(), InternalError> { Ok(self.engine.disable_tags(&[tag.to_str()?])) }().into()
+    }
+
+    pub fn tag_exists(&self, key: &CxxString) -> BoolResult {
+        || -> Result<bool, InternalError> { Ok(self.engine.tag_exists(key.to_str()?)) }().into()
+    }
+
+    pub fn matches(
+        &self,
+        url: &CxxString,
+        hostname: &CxxString,
+        source_hostname: &CxxString,
+        request_type: &CxxString,
+        third_party_request: bool,
+        previously_matched_rule: bool,
+        force_check_exceptions: bool,
+    ) -> BlockerResultResult {
+        || -> Result<BlockerResult, InternalError> {
+            let inner_blocker_result = self.engine.check_network_urls_with_hostnames_subset(
+                url.to_str()?,
+                hostname.to_str()?,
+                source_hostname.to_str()?,
+                request_type.to_str()?,
+                Some(third_party_request),
+                previously_matched_rule,
+                force_check_exceptions,
+            );
+            Ok(inner_blocker_result.into())
+        }()
+        .into()
+    }
+
+    pub fn get_csp_directives(
+        &self,
+        url: &CxxString,
+        hostname: &CxxString,
+        source_hostname: &CxxString,
+        request_type: &CxxString,
+        third_party_request: bool,
+    ) -> StringResult {
```

Comment:
Same arguments as `matches`, so [this](https://github.com/brave/brave-core/pull/17368#discussion_r1124040292) explanation applies equally and we can return the string directly.